### PR TITLE
[TOPI] Add more ops

### PR DESCRIPTION
**Pull request by nhynes**
This PR adds:
* `tvm::prod`
* `Tensor::numel`
* `topi::{mean, var}`

### CODE REVIEW

File: include/tvm/tensor.h [hunk 1]
```diff
@@ -47,6 +47,8 @@ class Tensor : public NodeRef {
   inline bool operator==(const Tensor& other) const;
   /*! \return The dimension of the tensor */
   inline size_t ndim() const;
+  /*! \return The number of elements in the tensor */
+  inline Expr numel() const;
```
Comment (tqchen) [inline include/tvm/tensor.h]: To make the API consistent with numpy, this function should be size

File: include/tvm/tensor.h [hunk 2]
```diff
@@ -173,6 +175,17 @@ inline size_t Tensor::ndim() const {
   return (*this)->shape.size();
 }
 
+inline Expr Tensor::numel() const {
+  Array<Expr> shape = (*this)->shape;
+  Type dtype = (*this)->dtype;
+  if (!shape.size()) return make_zero(dtype);
```
Comment (tqchen) [inline include/tvm/tensor.h]: zero dimensional array have size one

File: include/tvm/tensor.h [hunk 3]
```diff
@@ -173,6 +175,17 @@ inline size_t Tensor::ndim() const {
...
+  Expr numel = make_one(dtype);
```
Comment (tqchen) [inline include/tvm/tensor.h]: We should make product with the type of shape, instead of the dtype

File: topi/src/topi.cc [hunk 4]
```diff
@@ -201,6 +201,16 @@ TVM_REGISTER_GLOBAL("topi.sum")
   *rv = topi::sum(args[0], ArrayOrInt(args[1]), args[2]);
   });
 
+TVM_REGISTER_GLOBAL("topi.mean")
+.set_body([](TVMArgs args, TVMRetValue *rv) {
```
Comment (tqchen) [inline topi/src/topi.cc]: please add testcases for this

**Comment by tqchen** [comment]
ping @nhynes can you act on the comments?

**Comment by nhynes** [comment]
> ping @nhynes can you act on the comments?

For sure, sorry. It's just been an endless fractal of effort getting resnet to train in NNVM. I'll attend to this first thing tomorrow.

**Comment by tqchen** [comment]
The fusion rule is mainly done in topi and the current topi schedule injective is not as smart to handle the case. Two possible solutions exist

- break mean into sum(marked as reduction), div(injective)
   - So fusor will know to fuse at sum
- Keep mean, mark it as reduction and manually write the schedule for mean

**Comment by nhynes** [comment]
> break mean into sum(marked as reduction), div(injective)

I tried this, but then the CUDA test complains about injective not being a valid schedule.

**Comment by tqchen** [comment]
schedule_injective so far only handles the sequence that is all injective. 
Schedule reduce is needed for handling reductions.
By break mean into sum and div, i mean do it in nnvm level, instead of the topi level.

**Comment by tqchen** [comment]
close due to inactive status, please open another PR if we want to bring some of the changes

**Pull request closed without merging**